app/service/rpc_service.ts

```diff
@@ -238,6 +229,102 @@ function lengthPrefixMessage(requestData: Uint8Array) {
   return new Uint8Array(frame);
 }
 
+interface StreamState {
+  buffer: Uint8Array;
+  bufferedLength: number;
+  messageExpectedLength: number;
+  leftovers: Uint8Array;
+}
+
+async function transformLengthPrefixedReaderStreamToProtoMessageBytes(
+  reader: ReadableStreamDefaultReader | undefined,
+  callback: (e: Error | null, b: Uint8Array) => void
+) {
+  let streamState: StreamState = {
+    buffer: new Uint8Array(),
+    bufferedLength: 0,
+    messageExpectedLength: 0,
+    leftovers: new Uint8Array(),
+  };
+  while (true) {
+    // Start with any leftovers from the last turn.
+    let value: Uint8Array | undefined = streamState.leftovers;
+
+    // If we don't have any leftovers, read at least 5 bytes from the stream so we have at least a length prefix.
+    if (streamState.leftovers.length == 0) {
+      value = await readAtLeastNBytes(value, reader, 5);
+    }
+
+    // If we don't have any bytes, the stream is done.
+    if (!value) {
+      return;
+    }
+
+    // If we don't have an expected message length, we can consume the length prefix to get it.
+    if (streamState.messageExpectedLength == 0) {
+      value = consumeLengthPrefix(streamState, value);
+    }
+
+    // Read the rest of the bytes into the buffer until it's full, saving any leftovers.
+    consumeMessageChunk(streamState, value);
+
+    // If we still haven't receive the full message, loop again and read more until we do.
+    if (streamState.bufferedLength < streamState.messageExpectedLength) {
+      continue;
+    }
```

Comment:
sorry I randomly thought about this after approving - it looks like there should be a `readAtLeastNBytes()` call in here somewhere, reading at least `messageExpectedLength - value.length` bytes, before consuming the message chunk? Since otherwise it looks like we wrap around and attempt to consume a length prefix again in the case where streamState.leftovers.length === 0?

Reply:
Trying to think through this, I don't know how all three of these could be true:
- `streamState.bufferedLength < streamState.messageExpectedLength` i.e. we haven't filled up our buffer so we continue
- `streamState.messageExpectedLength == 0` so we consume another length prefix 

You mentioned this can happen if `streamState.leftovers.length == 0`, but I don't know how that can happen if `streamState.bufferedLength < streamState.messageExpectedLength`.

I totally believe that this is possible and you're right (maybe after multiple loops or something), I'm just trying to load the correct situation into my head so I can fix it properly.

Reply:
Would the case you're thinking about be handled if instead of checking that `streamState.leftovers.length == 0` we checked if we had less than 5 bytes of leftovers i.e.:
```
    // If we don't have any leftovers, read at least 5 bytes from the stream so we have at least a length prefix.
    if (streamState.leftovers.length < 5) {
      value = await readAtLeastNBytes(value, reader, 5);
    }
```

Reply:
Yeah that'd work, but I think it could be simplified because reading the length prefix and reading the message payload are both very similar - we want to read at least N bytes and then do something with the N bytes that we read. But currently it's a little bit hard to follow the control flow because we're using the nice `readAtLeastNBytes()` func for the length prefix, but not for the message payload.

I don't feel super strongly about it though since this code seems unlikely to be modified and the complexity is pretty self-contained.

Reply:
Took another crack, let me know what you think